### PR TITLE
[ndb] Add is_peer() method for interface objects

### LINKED ISSUE #835: Infinite loop when updating veth pair in netns
I am experiencing an infinite loop:
```
>>> from pyroute2 import NDB
>>> n = NDB()
>>> n.sources.add(target='foo', netns='foo-netns')
[running] <NetNS {'target': 'foo', 'netns': b'/var/run/netns/foo-netns'}>
>>> n.sources.add(target='bar', netns='cni-321b40ba-c240-c58c-6c74-0f82905d98ab')
[running] <NetNS {'target': 'bar', 'netns': b'/var/run/netns/cni-321b40ba-c240-c58c-6c74-0f82905d98ab'}>
>>> n.interfaces.create(ifname='foo0', kind='veth', target='foo', peer={ 'ifname': 'bar0', 'net_ns_fd': 'cni-321b40ba-c240-c58c-6c74-0f82905d98ab' }).commit()
{'flags': 4098, 'state': 'down', 'ifname': 'foo0', 'kind': 'veth', 'target': 'foo', 'peer': {'ifname': 'bar0', 'net_ns_fd': 'cni-321b40ba-c240-c58c-6c74-0f82905d98ab'}, 'tflags': 0, 'family': 0, 'ifi_type': 1, 'index': 2, 'change': 4294967295, 'address': '5a:53:96:4d:a2:fe', 'broadcast': 'ff:ff:ff:ff:ff:ff', 'mtu': 1500, 'link': 19, 'qdisc': 'noop', 'master': None, 'txqlen': 1000, 'linkmode': 0, 'net_ns_pid': None, 'ifalias': None, 'num_vf': None, 'group': 0, 'ext_mask': None, 'promiscuity': 0, 'num_tx_queues': 1, 'num_rx_queues': 1, 'carrier': 0, 'carrier_changes': 1, 'link_netnsid': 0, 'phys_port_name': None, 'proto_down': 0, 'gso_max_segs': 65535, 'gso_max_size': 65536, 'if_netnsid': None, 'carrier_up_count': 0, 'carrier_down_count': 1, 'new_ifindex': None, 'min_mtu': 68, 'max_mtu': 65535, 'alt_ifname': None, 'slave_kind': None}
>>> with n.interfaces[{ 'target': 'foo', 'ifname': 'foo0' }] as iface:
...   iface['state'] = 'up'

```

```
foo-netns # ip l
1: lo: <LOOPBACK> mtu 65536 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: foo0@if19: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 5a:53:96:4d:a2:fe brd ff:ff:ff:ff:ff:ff link-netns cni-321b40ba-c240-c58c-6c74-0f82905d98ab
```

```
cni-netns # ip l
1: lo: <LOOPBACK> mtu 65536 qdisc noqueue state DOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
3: eth0@if182: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1340 qdisc noqueue state UP
    link/ether 0e:8f:52:55:7b:ae brd ff:ff:ff:ff:ff:ff
19: bar0@if2: <BROADCAST,MULTICAST,M-DOWN> mtu 1500 qdisc noop state DOWN qlen 1000
    link/ether da:99:45:05:1a:c1 brd ff:ff:ff:ff:ff:ff
```

Note that in the default target/netns I have a vlan whose ifindex happens to match that of the `foo0` interface:
```
base-netns # ip l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: enp196s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether xx:xx:xx:xx:xx:xx brd ff:ff:ff:ff:ff:ff
3: staging0@enp196s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether xx:xx:xx:xx:xx:xx brd ff:ff:ff:ff:ff:ff
```

https://github.com/svinota/pyroute2/blob/1f8f9aa1cf541135ea9a7ea570056b1208f5fb41/pyroute2.ndb/pr2modules/ndb/objects/interface.py#L629-L640

This comparison seems intended to prevent such an infinite loop:
```python
if self.get('link') != spec['index']:
```

However, it is only partially valid. The conditional does not account for the `link_netnsid`. In other words, two interfaces returned via the `.getmany()` request may have the same `link` and yet point to completely different peers because those peers exist in separate namespaces.

### PR DESCRIPTION
**Pull request by centromere**
This determines whether a given interface "points to" another. It is relevant to vlans and veths.

This also fixes #835.

### CODE REVIEW

**Comment by svinota** [comment]
Thanks!

File: pyroute2.ndb/pr2modules/ndb/objects/interface.py [hunk 1]
```diff
@@ -613,6 +613,30 @@ def complete_key(self, key):
...
+    def is_peer(self, other):
+        '''Evaluate whether the given interface "points at" this one.'''
+        other_link = other.get('link')
+
+        if other_link != self['index']:
+            return False
+
+        other_link_netnsid = other.get('link_netnsid')
+        if other_link_netnsid is not None:
+            self_source = self.sources[self['target']]
```
Comment (svinota) [inline pyroute2.ndb/pr2modules/ndb/objects/interface.py:625]: This code block doesn't work as expected.

Assume we have three namespaces: main, ns1, ns2
And two veth pairs: main:veth1 + ns1:veth1p, main:veth2 + ns2:veth2p

Now on the kernel 5.11 I get following values for `link_netnsid`:

```
main (!netns)
  veth1['link_netnsid'] == 0
  veth2['link_netnsid'] == 0
ns1 (netns='/var/run/netns/ns1')
  veth1p['link_netnsid'] == 0
ns2 (netns='/var/run/netns/ns2')
  veth2p['link_netnsid'] == 1
```

Why? No idea. Probably it works better on later kernel, but a method used in the core functionality like rollbacks should work on older kernels as well.

I really like the idea of the method, and will continue to debug. @centromere if you have any ideas on that — pls provide comments or PR:s

I replace the method with a stub code that simply compares peer['link'] to host['index'], but later we should get it working properly.

Comment (centromere) [reply pyroute2.ndb/pr2modules/ndb/objects/interface.py:625]: Namespace IDs are themselves namespaced. There is a way to discover the list of NSIDs in another namespace. See [`get_netnsid_from_netnsid`](https://github.com/shemminger/iproute2/blob/9f366536edb5158343152604e82b968be46dbf26/ip/ipnetns.c#L375).

```
# ip netns list-id target-nsid 135
nsid 0 current-nsid unassigned (iproute2 netns name: cni-8f68967f-e070-1ee1-9e12-5e36162d84bb)
```

Comment (centromere) [reply pyroute2.ndb/pr2modules/ndb/objects/interface.py:625]: Kernel version 5.4:

```
# ip netns add ns1
# ip netns add ns2
# ip l add veth1 type veth peer veth1p netns ns1
# ip l add veth2 type veth peer veth2p netns ns2
```

```
>>> veth1 = n.interfaces['veth1']
>>> veth2 = n.interfaces['veth2']
>>> veth1p = n.interfaces[{ 'target': 'ns1', 'ifname': 'veth1p' }]
>>> veth2p = n.interfaces[{ 'target': 'ns2', 'ifname': 'veth2p' }]
>>> veth1['link_netnsid']
140
>>> veth2['link_netnsid']
141
>>> veth1p['link_netnsid']
0
>>> veth2p['link_netnsid']
0
>>> veth2p.is_peer(veth2)
True
>>> veth2.is_peer(veth2p)
False
```

This logic is at fault:
https://github.com/centromere/pyroute2/blob/1d07b07bc365c82255b99fe0e44d36ce762a5a9f/pyroute2.ndb/pr2modules/ndb/objects/interface.py#L625-L629

Comment (centromere) [reply pyroute2.ndb/pr2modules/ndb/objects/interface.py:625]: Here is an [additional resource](https://unix.stackexchange.com/questions/441876/how-to-find-the-network-namespace-of-a-veth-peer-ifindex).

**Pull request merged**